### PR TITLE
fix(router): Ensure APP_INITIALIZER of enabledBlocking option completes

### LINKED ISSUE #44355: Router initializer never completes if initial navigation fails
### Which @angular/* package(s) are relevant/releated to the feature request?

router

### Description

Rendering an app in SSR (using nguniversal) with an invalid URL (e.g. http://localhost:4200/electronics-spa/en/USD/Brands/Canon/c/brand_10%20or%20(1,2)=(select*from(select%20name_const(CHAR(82,88,106,99,113,78,74,70,73,118,87),1),name_const(CHAR(82,88,106,99,113,78,74,70,73,118,87),1))a)%20--%20and%201%3D1) causes the render to hang, and never release the taken resources.

The reason is that the navigation never completes on error, when the `Router` option `initialNavigation` is set to `enabled`.

### Proposed solution

Expose `RouterInitializer`.
EDIT: or make the router complete the navigation on error, if that makes sense.

### Alternatives considered

We came up with this _workaround_, but it would be nice if `ɵangular_packages_router_router_h` could be exposed.

```ts
import { Injectable, Injector, Provider } from '@angular/core';
import {
  NavigationError,
  Router,
  ROUTER_CONFIGURATION,
  ɵangular_packages_router_router_h as RouterInitializer,
} from '@angular/router';
import { filter, take } from 'rxjs/operators';

@Injectable()
export class CustomRouterInitializer extends RouterInitializer {
  appInitializer(): Promise<any> {
    const injector: Injector = (this as any).injector;
    const opts = injector.get(ROUTER_CONFIGURATION);

    if (
      opts.initialNavigation ===
      'enabled' /* for Ng11: || opts.initialNavigation === 'enabledBlocking' */
    ) {
      let resolve: any;
      const res = new Promise((r) => (resolve = r));

      injector
        .get(Router)
        .events.pipe(
          // We want to explicitly react on NavigationError event occurs just
          // after the first navigation start event
          take(2),
          filter((event) => event instanceof NavigationError)
        )
        .subscribe(() => resolve(false));

      super.appInitializer().then((val) => resolve(val));
      return res;
    }

    return super.appInitializer();
  }
}

export const routerFixProvider: Provider = {
  provide: RouterInitializer,
  useClass: CustomRouterInitializer,
};
```

The above workaround is for ng12. 
I noticed that in ng13 there's no similar private API available, unless I'm missing something.

### PR DESCRIPTION
**Pull request by atscott**
Previously, if `initialNavigation` were set to `enabledBlocking`, the
Router's `APP_INITIALIZER` would never resolve if that initial
navigation failed. This results in the application load hanging and
never completing.

fixes https://github.com/angular/angular/issues/44355

### CODE REVIEW

**Comment by atscott** [comment]
[presubmit](https://fusion2.corp.google.com/presubmit/tap/449417433/OCL:449417433:BASE:449417472:1652861282140:75f7e103/targets)

File: packages/router/src/router_module.ts [hunk 1]
```diff
@@ -601,7 +607,41 @@ export class RouterInitializer implements OnDestroy {
...
+                  if (e instanceof NavigationEnd) {
+                    // Navigation succeeded if we get to `NavigationEnd`
+                    return true;
+                  }
+                  const newNavigationStarted =
+                      !!router.getCurrentNavigation() && router.getCurrentNavigation()?.id !== e.id;
+                  // TODO(atscott): Do not rely on the string reason to determine if cancelation is
+                  // redirecting
+                  const redirectingWithUrlTree = e instanceof NavigationCancel ?
+                      e.reason.indexOf(REDIRECTING_CANCELATION_REASON) !== -1 :
```
Comment (atscott) [inline packages/router/src/router_module.ts:629]: note to reviewer: There's no other information that indicates the cancelation was due to a guard returning a `UrlTree`. You can see here that the new navigation also happens _after_ the `NavigationCancel` event so we also can't rely on navigation IDs to check if it's redirecting:
https://github.com/angular/angular/blob/e92799c97e9aa7fefbb40560f932994150fdb17f/packages/router/src/router.ts#L983-L1018

Not relying on the "Redirecting to" string would require a new property on `NavigationCancel`, which would need to be considered a feature (so we couldn't land this fix in a patch).

**Comment by atscott** [comment]
@alxhub This is the first step for fixing #16211. This is purely the bug fix side of it that can land in a patch and then we can finalize what we want to expose as the feature option for a fallback navigation if the `initialNavigation` fails.

File: packages/router/src/router_module.ts [hunk 2]
```diff
@@ -601,7 +607,40 @@ export class RouterInitializer implements OnDestroy {
...
+                filter(
+                    (e): e is NavigationEnd|NavigationCancel|NavigationError =>
+                        e instanceof NavigationEnd || e instanceof NavigationCancel ||
+                        e instanceof NavigationError),
+                map(e => {
+                  if (e instanceof NavigationEnd) {
+                    // Navigation succeeded if we get to `NavigationEnd`
+                    return true;
+                  }
+                  const newNavigationStarted = router.navigationId !== e.id;
```
Comment (atscott) [inline packages/router/src/router_module.ts:624]: This `navigationId` was made `@internal` to be read here and is updated before the transition for the new navigation is pushed to the pipe that has the `switchMap`. Another option would be to look at the error string again like we had to do with the redirecting via `UrlTree` (https://github.com/angular/angular/blob/e92799c97e9aa7fefbb40560f932994150fdb17f/packages/router/src/router.ts#L958-L960)
A third option would be to move the update of `currentNavigation` to before the `switchMap` to ensure it's updated before the `finalize` block runs and emits the `NavigationCancel` event
https://github.com/angular/angular/blob/e92799c97e9aa7fefbb40560f932994150fdb17f/packages/router/src/router.ts#L656-L667
I'd be a bit more worried about this due to general fragility around timing.

**Review by jessicajaniuk — APPROVED** [review]
reviewed-for: size-tracking

File: packages/router/src/shared.ts [hunk 3]
```diff
@@ -111,6 +111,7 @@ export function convertToParamMap(params: Params): ParamMap {
   return new ParamsAsMap(params);
 }
 
+export const REDIRECTING_CANCELATION_REASON = 'Redirecting to ';
```
Comment (alxhub) [inline packages/router/src/shared.ts:114]:
```suggestion
export const REDIRECTING_CANCELLATION_REASON = 'Redirecting to ';
```

**Review by alxhub — APPROVED** [review]

**Review by alxhub — APPROVED** [review]
Reviewed-for: size-tracking

**Comment by atscott** [comment]
This PR was merged into the repository by commit e8c7dd10e93d338456131e11d600bf758ee48634.

**Pull request closed without merging**